### ISSUE TITLE
🐞 Parallel Dagger clients with overlapping LLB vertices can receive the wrong progress logs

### ISSUE THREAD

**Issue by sipsma** [dagger/dagger#2283]
### What is the issue?

A dagger client running in parallel with another one (i.e. multiple invocations of `dagger do` that are talking to the same buildkitd at the same time) can see logs meant for the other client if the two share any LLB vertices with equal digests (which includes both the vertex itself and all its ancestors).

This is a bug in Buildkit where it deduplicates the execution of exactly equivalent vertices, but in doing so also deduplicates any metadata associated with those vertices. The metadata may not be the same because metadata does not contribute to the digest of the vertex. Dagger uses the vertex name metadata to format its progress logs, so this can cause very confusing output.

For example:
```
package testing

import (
	"dagger.io/dagger"
	"dagger.io/dagger/core"
)

dagger.#Plan & {
	actions: {
		base: core.#Pull & {
			source: "alpine:3.15"
		}

		verifyA: core.#Exec & {
			input: base.output
			args: ["echo", "HI"]
			always: true
		}

		verifyB: core.#Exec & {
			input: base.output
			args: ["echo", "HI"]
			always: true
		}
	}
}
```

With that plan at `test.cue`, you can run the following:
```
dagger do -p test.cue --log-format=plain verifyA & dagger do -p test.cue --log-format=plain verifyB >/dev/null 2>&1
```

This results in `do verifyA` to run and print to the output in parallel with `verifyB`, whose log output is hidden. Quite often (but not every time), you will see that `do verifyA` shows output like:
```
4:00PM INF actions.verifyA | computing
4:00PM INF actions.verifyA | completed    duration=100ms
4:00PM INF actions.verifyB | #3 0.088 HI
```

Which incorrectly includes `actions.verifyB`. 

This needs to be addressed upstream in Buildkit of course. As for any possible solutions in the meantime, I can't think of any right away but will give it some more thought.

### Dagger version

dagger devel (9be4d4d8) linux/arm64

### OS version

Linux buildkit_dev 5.16.0-6-arm64 #1 SMP Debian 5.16.18-1 (2022-03-29) aarch64 GNU/Linux

**Comment by sipsma** [comment]
Issue in Buildkit: https://github.com/moby/buildkit/issues/2820

**Comment by dolanor** [comment]
I guess, we could try to figure out which action should be part of the action run. And filter everything that is not part of the actions for this run.
I don't know how feasible it is, though.

**Comment by sipsma** [comment]
> I guess, we could try to figure out which action should be part of the action run. And filter everything that is not part of the actions for this run.

I think this just needs to be addressed upstream in buildkit. It's probably not worth trying to find a hack workaround for us here

**Comment by shykes** [comment]
Is this still relevant?

**Comment by sipsma** [comment]
> Is this still relevant?

Yes, this will impact the use of progress groups in the new logger implementation @aluzzardi is working on, due to the fact that progress groups are part of llb metadata

**Comment by sipsma** [comment]
Fairly certain this will no longer be an issue after moving to OTEL, so closing out since that's coming soon and this issue is ancient
